contrib/twitchtv/twirp/twirp.go

```diff
@@ -0,0 +1,208 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// Package twirp provides tracing functions for tracing clients and servers generated
+// by the twirp framework (https://github.com/twitchtv/twirp).
+package twirp // import "gopkg.in/DataDog/dd-trace-go.v1/contrib/twitchtv/twirp"
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/twitchtv/twirp"
+)
+
+type contextKey int
+
+const (
+	twirpErrorKey contextKey = iota
+)
+
+// HTTPClient is duplicated from twirp's generated service code.
+// It is declared in this package so that the client can be wrapped
+// to initiate traces.
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type wrappedClient struct {
+	c   HTTPClient
+	cfg *config
+}
+
+// WrapClient wraps a HTTPClient to add disributed tracing to its requests.
+func WrapClient(c HTTPClient, opts ...Option) HTTPClient {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &wrappedClient{c: c, cfg: cfg}
+}
+
+func (wc *wrappedClient) Do(req *http.Request) (*http.Response, error) {
+	opts := []tracer.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeHTTP),
+		tracer.Tag(ext.HTTPMethod, req.Method),
+		tracer.Tag(ext.HTTPURL, req.URL.Path),
+	}
+	if wc.cfg.serviceName != "" {
+		opts = append(opts, tracer.ServiceName(wc.cfg.serviceName))
+	}
+	ctx := req.Context()
+	if pkg, ok := twirp.PackageName(ctx); ok {
+		opts = append(opts, tracer.Tag("twirp.package", pkg))
+	} else {
+		fmt.Println("no package")
+	}
+	if svc, ok := twirp.ServiceName(ctx); ok {
+		opts = append(opts, tracer.Tag("twirp.service", svc))
+	}
+	if method, ok := twirp.MethodName(ctx); ok {
+		opts = append(opts, tracer.Tag("twirp.method", method))
+	}
+	if !math.IsNaN(wc.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, wc.cfg.analyticsRate))
+	}
+	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Header)); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+
+	span, ctx := tracer.StartSpanFromContext(req.Context(), "twirp.request", opts...)
+	defer span.Finish()
+
+	req = req.WithContext(ctx)
+	err := tracer.Inject(span.Context(), tracer.HTTPHeadersCarrier(req.Header))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "contrib/twitchtv/twirp.wrappedClient: failed to inject http headers: %v\n", err)
```

Comment:
Why are we not using the internal logging package?

Reply:
Updated to use internal logging package.